Core/tests/Unit/ExponentialBackoffTest.php

```diff
@@ -160,4 +160,58 @@ public function delayProvider()
             [10, 60000000, 60000000]
         ];
     }
+
+    public function testOnRetryExceptionFunction()
+    {
+        $args = ['foo' => 'bar'];
+        $onRetryExceptionFunction = function (
+            $ex,
+            $retryAttempt,
+            $arguments
+        ) {
+            $message = sprintf('%s, %s', $retryAttempt, $arguments[0]['foo']);
+            throw new \Exception($message);
+        };
+
+        $result = '';
+        $expected = '0, bar';
+        $backoff = new ExponentialBackoff(null, null, $onRetryExceptionFunction);
+        try {
+            $backoff->execute(
+                function ($options) use ($args) {
+                    throw new \Exception();
+                },
+                [$args]
+            );
+        } catch (\Exception $ex) {
+            $result = $ex->getMessage();
+        }
+
+        $this->assertEquals($expected, $result);
+    }
+
+    public function testOnExecutionStartFunction()
+    {
+        $args = ['foo' => 'bar'];
+        $onExecutionStartFunction = function ($arguments) {
+            $message = sprintf('%s', $arguments[0]['foo']);
+            throw new \Exception($message);
+        };
+
+        $result = '';
+        $expected = 'bar';
+        $backoff = new ExponentialBackoff(null, null, null, $onExecutionStartFunction);
+        try {
+            $backoff->execute(
+                function ($options) use ($args) {
+                    throw new \Exception();
```

Comment:
Just like above, but here we can also put a fail() inside anonymous function passed to execute() to denote that it will never reach due to the exception inside `onExecutionStart`

Reply:
I didn't get this :(, can you please elaborate?

Reply:
So, basically I was suggesting to use `$this->fail()` instead of the exception statement.

So, here we can use `self::fail("Reached execute code block, but should have failed in onExecutionStartFunction")` or some other verbose statement 😄

Reply:
Yeah, this seems better.

Applied in commit [ba03c51](https://github.com/googleapis/google-cloud-php/pull/5953/commits/ba03c51d90bb70cd400d755623229350d97ccd2d).